File: DS/Stacks/Java/reverse_linked_list.java
import java.util.Stack;

public class Linkedlist {
    Node head;

    static class Node
    {
        int data;
        Node link;

        Node(int num){
            data = num;
            link = null;
        }
    }

    //insert a node at the end of linked list
    public void append(int num){
        Node temp = new Node(num);
        if(head == null){
            head = temp;
            return;
        }
        Node temp1 = head;
        while (temp1.link!=null){
            temp1 = temp1.link;
        }

        temp1.link = temp;
    }

    public void reverse() {
        Node temp = head;
        Stack<Node> stack = new Stack<Node>();
        while (temp != null) {
            stack.push(temp);
            temp = temp.link;
        }
        temp = stack.peek();
        head = temp;
        stack.pop();

        while (!stack.empty()){
            temp.link = stack.peek();
            temp = temp.link;
            stack.pop();
        }
        temp.link = null;
    }

    //print linked list
    public void print(){
        Node temp = head;
        while (temp!=null){
            System.out.print(temp.data + " ");
            temp = temp.link;
        }
    }

    //main
    public static void main(String[] args){
        Linkedlist list = new Linkedlist();
        list.append(1);
        list.append(2);
        list.append(3);
        list.append(4);
        list.append(5);

        list.reverse();

        list.print();
    }
}

// output: 5 4 3 2 1

File: DS/Stacks/Python/reverse_linked_list.py
class Node:
    def __init__(self, num):
        self.data = num
        self.link = None


class LinkedList:

    def __init__(self):
        self.head = None

    # insert node at the end of linked list
    def append(self, num):
        temp = Node(num)
        if self.head is None:
            self.head = temp
            return

        temp1 = self.head
        while temp1.link is not None:
            temp1 = temp1.link

        temp1.link = temp

    def reverse(self):
        stack = []
        temp = self.head
        while temp is not None:
            stack.append(temp)
            temp = temp.link

        temp = stack.pop()
        self.head = temp

        while len(stack) != 0:
            temp.link = stack.pop()
            temp = temp.link

        temp.link = None

    # print
    def print_list(self):
        temp = self.head
        while temp:
            print(temp.data, end=" ")
            temp = temp.link


if __name__ == '__main__':
    l_list = LinkedList()

    l_list.append(1)
    l_list.append(2)
    l_list.append(3)
    l_list.append(4)
    l_list.append(5)

    l_list.reverse()

    l_list.print_list()

# output: 5 4 3 2 1
